File: src/main/scala/org/tindalos/principle/domain/analyzers/submodulesblueprint/SubmodulesBlueprintAnalyzer.scala
package org.tindalos.principle.domain.analyzers.submodulesblueprint

import org.tindalos.principle.domain.AnalysisInput
import org.tindalos.principle.domain.analyzers.Analyzer
import org.tindalos.principle.domain.constraints.Constraints
import org.tindalos.principle.domain.core.{Package, PackageStructureBuilder}
import org.tindalos.principle.domain.core.packages.PackageWithMetrics

import scala.collection.JavaConverters.asScalaSetConverter
import scala.collection.JavaConverters._

class SubmodulesBuilder(packageStructureBuilder: PackageStructureBuilder) {
  def build(submoduleDefinitions: SubmoduleDefinitions, packages: List[PackageWithMetrics], basePackageName: String): Set[Submodule] = {

    submoduleDefinitions.checkNoOverlaps()
    val basePackage = packageStructureBuilder.build(packages.asInstanceOf[List[Package]], basePackageName)


    def convert(submoduleDefinition: SubmoduleDefinition): Submodule = {
      val packages = submoduleDefinition.packages().asScala.map(reference => basePackage.toMap().get(reference) match {
        case None => throw new InvalidBlueprintDefinitionException("Package does not exist: " + reference)
        case Some(aPackage) => aPackage
      })
      new Submodule(submoduleDefinition.id, packages.toSet.asInstanceOf[Set[PackageWithMetrics]].asJava, submoduleDefinition.getLegalDependencies.asScala.toSet.asJava)
    }

    submoduleDefinitions.getDefinitions().asScala.values.map(convert).toSet

  }
}

class SubmodulesBlueprintAnalyzer(submodulesBuilder: SubmodulesBuilder) extends Analyzer {

  override def isEnabled(designQualityChecks: Constraints) = designQualityChecks.submoduleDefinitions().isPresent

  override def analyze(checkInput: AnalysisInput): SubmodulesBlueprintAnalysisResult =

    if (checkInput.submoduleDefinitions().isPresent) {
      val submoduleDefinitions = checkInput.submoduleDefinitions().get()

        try {
          val submodules = submodulesBuilder.build(
            submoduleDefinitions,
            checkInput.packages, checkInput.analysisPlan.basePackage)

          val (aID, aMD) = problematicDependencies(submodules)
          SubmodulesBlueprintAnalysisResult.withViolations(submoduleDefinitions.violationThreshold, aID, aMD)
        }
        catch {
          case ex: OverlappingSubmoduleDefinitionsException =>
            SubmodulesBlueprintAnalysisResult.withOverlaps(submoduleDefinitions.violationThreshold, ex.getOverlaps().asScala.toSet.asJava)
        }
    } else {
      SubmodulesBlueprintAnalysisResult.empty(0)
    }

  private def problematicDependencies(submodules: Set[Submodule]): (java.util.Map[Submodule, java.util.Set[Submodule]], java.util.Map[Submodule, java.util.Set[Submodule]]) = {
    val emptyMap = Map[Submodule, Set[Submodule]]()
    val (aID, aMD) = submodules.foldLeft((emptyMap, emptyMap))((acc, submodule) => {
      val otherSubmodules = submodules.filterNot(_.equals(submodule))
      val illegalDependencies = submodule.findIllegalDependencies(otherSubmodules.asJava).asScala.toSet
      val missingDependencies = submodule.findMissingPredefinedDependencies(otherSubmodules.asJava).asScala.toSet

      val aID2 = if (illegalDependencies.isEmpty) acc._1 else acc._1 + (submodule -> illegalDependencies)
      val aMD2 = if (missingDependencies.isEmpty) acc._2 else acc._2 + (submodule -> missingDependencies)

      (aID2, aMD2)
    })
    (aID.map { case (k, v) => k -> v.asJava }.asJava, aMD.map { case (k, v) => k -> v.asJava }.asJava)
  }
}

File: src/main/java/org/tindalos/guardrails/internal/domain/analyzers/submodulesblueprint/SubmodulesBlueprintAnalyzer.java
package org.tindalos.guardrails.internal.domain.analyzers.submodulesblueprint;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

import org.tindalos.guardrails.internal.domain.analyzers.Analyzer;
import org.tindalos.guardrails.internal.domain.constraints.Constraints;
import org.tindalos.guardrails.internal.domain.constraints.submodules.OverlappingSubmoduleDefinitionsException;
import org.tindalos.guardrails.internal.domain.plan.AnalysisInput;

/**
 * Analyzes discovered packages against declared submodule blueprint definitions.
 */
public class SubmodulesBlueprintAnalyzer implements Analyzer {

    private final SubmodulesBuilder submodulesBuilder;

    public SubmodulesBlueprintAnalyzer(SubmodulesBuilder submodulesBuilder) {
        this.submodulesBuilder = submodulesBuilder;
    }

    @Override
    public boolean isEnabled(Constraints designQualityConstraints) {
        return designQualityConstraints.submoduleDefinitions().isPresent();
    }

    @Override
    public SubmodulesBlueprintAnalysisResult analyze(AnalysisInput checkInput) {
        if (checkInput.submoduleDefinitions().isEmpty()) {
            return SubmodulesBlueprintAnalysisResult.empty(0);
        }

        var submoduleDefinitions = checkInput.submoduleDefinitions().orElseThrow();
        try {
            Set<Submodule> submodules = Set.copyOf(submodulesBuilder.build(
                submoduleDefinitions,
                checkInput.packages(),
                checkInput.analysisPlan().basePackage()));

            var dependencyMaps = problematicDependencies(submodules);
            return SubmodulesBlueprintAnalysisResult.withViolations(
                submoduleDefinitions.violationThreshold(),
                dependencyMaps.illegalDependencies(),
                dependencyMaps.missingDependencies());
        } catch (OverlappingSubmoduleDefinitionsException ex) {
            return SubmodulesBlueprintAnalysisResult.withOverlaps(
                submoduleDefinitions.violationThreshold(),
                ex.getOverlaps());
        }
    }

    private DependencyMaps problematicDependencies(Set<Submodule> submodules) {
        Map<Submodule, Set<Submodule>> illegalDependenciesBySubmodule = new HashMap<>();
        Map<Submodule, Set<Submodule>> missingDependenciesBySubmodule = new HashMap<>();

        for (Submodule submodule : submodules) {
            Set<Submodule> otherSubmodules = new HashSet<>(submodules);
            otherSubmodules.remove(submodule);

            Set<Submodule> illegalDependencies = Set.copyOf(submodule.findIllegalDependencies(otherSubmodules));
            Set<Submodule> missingDependencies = Set.copyOf(submodule.findMissingPredefinedDependencies(otherSubmodules));

            if (!illegalDependencies.isEmpty()) {
                illegalDependenciesBySubmodule.put(submodule, illegalDependencies);
            }
            if (!missingDependencies.isEmpty()) {
                missingDependenciesBySubmodule.put(submodule, missingDependencies);
            }
        }

        return new DependencyMaps(Map.copyOf(illegalDependenciesBySubmodule), Map.copyOf(missingDependenciesBySubmodule));
    }

    /**
     * Internal immutable pair of dependency maps.
     */
    private record DependencyMaps(
        Map<Submodule, Set<Submodule>> illegalDependencies,
        Map<Submodule, Set<Submodule>> missingDependencies) {
    }
}